File: lib/graph/python/topological_sort.py
def topological_sort(g: list[list[int]]):
    n = len(g)
    in_deg = [0]*n
    for v in range(n):
        for u in g[v]:
            in_deg[u] += 1
    
    stack = [v for v in range(n) if in_deg[v] == 0]
    order = []
    while stack:
        v = stack.pop()
        order.append(v)
        for u in g[v]:
            in_deg[u] -= 1
            if in_deg[u] == 0:
                stack.append(u)

    for v in range(n):
        if in_deg[v] != 0:
            return False

    return order


from sys import setrecursionlimit
setrecursionlimit(2**30)

def topological_sort(g: list[list[int]]):
    n = len(g)
    # identify roots
    roots = set(range(n))
    for i in range(n):
        for j in g[i]:
            roots.discard(j)
    
    if not roots:
        return False # not a dag

    topo_order = []
    visited = [False]*n
    rec_visited = [False]*n
    def dfs(v):
        if rec_visited[v]:
            return False # not a dag
        if not visited[v]:
            visited[v] = True
            rec_visited[v] = True

            for u in g[v]:
                if not dfs(u):
                    return False

            rec_visited[v] = False
            topo_order.append(v)
        return True

    for root in roots:
        if not dfs(root):
            return False
    topo_order.reverse()
    return topo_order

File: lib/graph/rust/topological_sort.rs
// NOT TESTED

fn topological_sort(g: &Vec<Vec<usize>>) -> Option<Vec<usize>> {
    let n = g.len();
    let mut in_deg = vec![0; n];
    for i in 0..n {
        for j in &g[i] {
            in_deg[*j] += 1;
        }
    }

    let mut roots = Vec::new();
    for i in 0..n {
        if in_deg[i] == 0 {
            roots.push(i);
        }
    }

    if roots.len() == 0 {
        return None;
    }

    let mut topo_order = Vec::new();
    let mut visited = vec![false; n];
    let mut rec_visited = vec![false; n];

    fn dfs(
        v: usize,
        g: &Vec<Vec<usize>>,
        rec_visited: &mut [bool],
        visited: &mut [bool],
        topo_order: &mut Vec<usize>,
    ) -> bool {
        if rec_visited[v] {
            return false;
        }
        if !visited[v] {
            visited[v] = true;
            rec_visited[v] = true;

            for u in &g[v] {
                if !dfs(*u, g, rec_visited, visited, topo_order) {
                    return false;
                }
            }
            rec_visited[v] = false;
            topo_order.push(v)
        }
        return true;
    }

    for root in roots {
        if !dfs(root, g, &mut rec_visited, &mut visited, &mut topo_order) {
            return None;
        }
    }
    topo_order.reverse();
    Some(topo_order)
}